### ISSUE TITLE
It would be cool to have dynamic modules support similar to Redis

### ISSUE THREAD

**Issue by WithoutCaps** [nats-io/nats-server#1479]
## Feature Request
It would be cool to have dynamic modules support similar to Redis
#### Use Case:
The community could write modules which would allow extending NATS functionality
#### Proposed Change:
Add API which would allow importing custom modules - plugins if you will
#### Who Benefits From The Change(s)?
Community
#### Alternative Approaches
Edit source

**Comment by ripienaar** [comment]
Do you have examples of the kind of plugin you'd write?

**Comment by WithoutCaps** [comment]
Yeah, I had this weird/dumb idea of integrating request/reply cache with Redis at NATS level. Something along the lines of such pipeline:
1. Request sent
2. NATS notices that the requested subject is cached
3. NATS looks up Redis
4. If record found - publishes it as a response
5. If record not found - continues off with other subscriptions

TL;DR: That's something along the lines of setting consumers priority/failed requests handling

**Comment by derekcollison** [comment]
We do have future plans to enhance the NATS network elements with secure WASM so this may be applicable at that point. Need WASM and more importantly WASI to progress a bit, also desire a non-cgo implementation for running WASM code inside the NATS server and of course utilizing our security model properly similar to NATS 2.0 Authentication and Authorization etc.
